src/tokens/ERC1155.vy

```diff
@@ -178,57 +270,207 @@ def balanceOfBatch(owners: DynArray[address, max_value(uint16)], ids: DynArray[u
+def set_uri(id: uint256, uri: String[432]):
     """
-    return empty(String[512])
+    @dev Sets token URI for a given `token_id`.
+    @notice This is decoupled from the `mint` function
+        since multiple of the same `token_id` may be
+        minted. However, permissions are shared with
+        `is_minter`. Only minters have authorization
+        to change token URIs.
+    @param token_id The 32-byte token identifier.
+    @param uri The maximum 432-character user-readable
+        string URI for computing `tokenURI`.
+    """
+    assert self.is_minter[msg.sender], "ERC1155: not authorized to set token URI"
+    self._token_uris[id] = uri
+    log URI(uri, id)
+
+
+@external
+def transfer_ownership(new_owner: address):
+    """
+    @dev Sourced from {Ownable-transfer_ownership}.
+    @notice See {Ownable-transfer_ownership} for
+            the function docstring.
+    """
+    assert msg.sender == self.owner, "Ownable: caller is not the owner"
+    assert new_owner != empty(address), "Ownable: new owner is the zero address"
+    old_owner: address = self.owner
+    self.owner = new_owner
+    log OwnershipTransferred(old_owner, new_owner)
+
+
+@external
+def renounce_ownership():
+    """
+    @dev Leaves the contract without an owner.
+    @notice Renouncing ownership will leave the
+            contract without an owner, thereby
+            removing any functionality that is
+            only available to the owner. Note
+            that the `owner` is also removed from
+            the list of allowed minters.
+
+            WARNING: All other existing `minter`
+            addresses will still be able to create
+            new tokens. Consider removing all non-owner
+            minter addresses first via `set_minter`
+            before calling `renounce_ownership`.
+    """
+    assert msg.sender == self.owner, "Ownable: caller is not the owner"
+    self.is_minter[msg.sender] = False
+    old_owner: address = self.owner
+    self.owner = empty(address)
+    log OwnershipTransferred(old_owner, empty(address))
+
+
+@external
+def set_minter(minter: address, status: bool):
+    """
+    @dev Adds or removes an address `minter` to/from the
+         list of allowed minters. Note that only the
+         `owner` can add or remove `minter` addresses.
+         Also, the `minter` cannot be the zero address.
+         Eventually, the `owner` cannot remove himself
+         from the list of allowed minters.
+    @param minter The 20-byte minter address.
+    @param status The Boolean variable that sets the status.
+    """
+    assert msg.sender == self.owner, "AccessControl: caller is not the owner"
+    assert minter != empty(address), "AccessControl: minter is the zero address"
+    assert minter != msg.sender, "AccessControl: minter is owner address"
+    self.is_minter[minter] = status
+    log RoleMinterChanged(minter, status)
+
+
+@external
+def safe_mint(owner: address, id: uint256, amount: uint256, data: Bytes[1024]):
```

Comment:
well IMHO we would need the following:
- an `external` function `safe_mint` that calls `_mint`
- an `internal` function `_mint`
- an `external` function `safe_mint_batch` that calls `_mint_batch`
- an `internal` function `_mint_batch`

Reply:
done!

Reply:
so in terms of composability, can we have the access control pattern on the `external` function level (like that some people can import the internal function(s) and allow for public minting etc. - the internal functions should without be access restriction.).

For instance, like that:
```vy
def safe_mint(owner: address, id: uint256, amount: uint256, data: Bytes[1024]):
    assert self.is_minter[msg.sender], "AccessControl: access is denied"
    self._mint(owner, id, amount, data)
``` 

Like I did here: https://github.com/pcaversaccio/snekmate/blob/main/src/tokens/ERC20.vy#L361 or here: https://github.com/pcaversaccio/snekmate/blob/main/src/tokens/ERC721.vy#L542.